<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Создаем-псевдослучайно-квадратную-матрицу" data-toc-modified-id="Создаем-псевдослучайно-квадратную-матрицу-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Создаем псевдослучайно квадратную матрицу</a></span></li><li><span><a href="#Проверяем,-что-созданная-матрица-обратимая-$P*P^{-1}-=-E$" data-toc-modified-id="Проверяем,-что-созданная-матрица-обратимая-$P*P^{-1}-=-E$-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Проверяем, что созданная матрица обратимая $P*P^{-1} = E$</a></span></li><li><span><a href="#Обучаем-модель,-получаем-метрику-на-необработанных-данных" data-toc-modified-id="Обучаем-модель,-получаем-метрику-на-необработанных-данных-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Обучаем модель, получаем метрику на необработанных данных</a></span></li><li><span><a href="#Умножаем-сет-признаков-на-обратимую-матрицу.-Обучаем-модель,-получаем-метрику-на-обработанных-данных." data-toc-modified-id="Умножаем-сет-признаков-на-обратимую-матрицу.-Обучаем-модель,-получаем-метрику-на-обработанных-данных.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Умножаем сет признаков на обратимую матрицу. Обучаем модель, получаем метрику на обработанных данных.</a></span></li><li><span><a href="#Сравнение-R2-показатели-модели" data-toc-modified-id="Сравнение-R2-показатели-модели-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Сравнение R2 показатели модели</a></span></li></ul></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод:</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

In [2]:
try:
    data = pd.read_csv('/datasets/insurance.csv')
except:
    data = pd.read_csv(r'C:\Users\vadim.pilipenko\Desktop\личное\Практикум\ML\прочее\insurance.csv')

display(data.head(10), data.shape)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

(5000, 5)

In [4]:
features = data.iloc[:,:-1]
target = data.iloc[:,-1]

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$


Свойства матриц используемые в рассчетах:
    
$A(BC) = (AB)C$
    
$(AB)^T = B^TA^T$
    
$(AB)^{-1} = B^{-1}A^{-1}$
    
$(ABC)^{-1} = C^{-1}B^{-1}A^{-1}$
    
$AA^{-1}=A^{-1}A=E$
    
**Ответ:** При умножении признаков на обратимую матрицу качество линейной регрессии не измениться.

**Обоснование:** 
Умножим матрицу $X$ на обратимую матрицу $P$ в формуле обучения вектора весов $w$:
$$ w' = ((XP)^T XP)^{-1} (XP)^T y$$
Используя свойства матриц проведем преобразование:

$$((XP)^T XP)^{-1} (XP)^T y =$$

$$(P^TX^T XP)^{-1} P^TX^T y =$$
    
$$(P^T(X^T X)P)^{-1} P^TX^T y =$$
    
$$P^{-1} (X^T X)^{-1} (P^T)^{-1} P^TX^T y =$$

$$P^{-1} (X^T X)^{-1} EX^T y =$$
    
$$P^{-1} (X^T X)^{-1} X^T y$$
    
Теперь вычислим предсказания $a'$ используя формулу $a' = XPw'$:

$$XPP^{-1}(X^TX)^{-1}X^T y =$$

$$XE(X^TX)^{-1}X^T y =$$

$$X(X^TX)^{-1}X^T y = Xw$$

**Получаем, что $a$ = $a'$, таким образом предсказания, а следовательно и качество линейной регрессии при умножении признаков на обратимую матрицу не меняется**


## Алгоритм преобразования

**Алгоритм**

1. Создаем псевдослучайно квадратную матрицу 
2. Проверяем, что созданная матрица обратимая
3. Обучаем модель, получаем метрику на необработанных данных.
4. Умножаем сет признаков на обратимую матрицу.
5. Обучаем модель, получаем метрику на обработанных данных.
5. Сравниваем показатели метрики.

## Проверка алгоритма

In [5]:
model=LinearRegression()

### Создаем псевдослучайно квадратную матрицу

In [6]:
random_matrix = np.random.randint(1,10,[features.shape[1],features.shape[1]]) #создадим псевдослучайную матрицу
print(random_matrix)


[[3 1 2 9]
 [6 5 2 5]
 [3 8 4 8]
 [5 7 5 6]]


### Проверяем, что созданная матрица обратимая $P*P^{-1} = E$

In [7]:
inv_random_matrix = np.linalg.inv(random_matrix) 
print(np.around(inv_random_matrix@random_matrix))#учтем, что точность вычислений  не будет идеальной, округлим.

[[ 1. -0. -0.  0.]
 [-0.  1.  0. -0.]
 [ 0.  0.  1. -0.]
 [ 0.  0.  0.  1.]]


Делим признаки и таргет на учебную и тренировочную выборки

In [8]:
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size = 0.25, random_state=12345)

### Обучаем модель, получаем метрику на необработанных данных

In [9]:
model.fit(features_train,target_train)
print(model.score(features_test,target_test))# Метод score в модели линейной регрессии возвращает коэффициент детерминации (R2)

0.435227571270266


### Умножаем сет признаков на обратимую матрицу. Обучаем модель, получаем метрику на обработанных данных.

In [10]:
features = features@random_matrix
features_train, features_test, target_train, target_test  = train_test_split(features, target, test_size = 0.25, random_state=12345)
model.fit(features_train,target_train)
print(model.score(features_test,target_test))

0.4352275712701702


### Сравнение R2 показатели модели
До умножения на обратимую матрицу R2 = 0.435227571270266 после умножения R2 = 0.4352275712702385. С высокой точностью они равны. 

## Вывод:
R2 модели линейной регрессии на данных без обработки и R2 после умножения на обратимую матрицу можно считать равными, что согласуется с приведенными выше математическиим расчетами.
Качество модели не изменяется.